### ISSUE TITLE
On Development Process Transparency & Roadmap

### ISSUE THREAD

**Issue by KelSolaar** [AcademySoftwareFoundation/OpenColorIO#1124]
Hello,

I'm starting an issue to centralise the discussions in relation to my two comments [here](https://github.com/AcademySoftwareFoundation/OpenColorIO/issues/1100#issuecomment-674340278) and [there](https://github.com/AcademySoftwareFoundation/OpenColorIO/pull/1102#issuecomment-682302942).

# Introduction

With PR #1102, OpenColorIO is about to receive three new grading transforms. I want to clarify right from the beginning, that the discussion in this issue is not about the transforms themselves but how we got there and how we can improve the process to get there because it raised some valid concerns in the industry.

The gist of the concerns being: How can 22K lines of code for a significant feature land in the repo without an issue or a public thread being created first to discuss the feature. Put another way: How it is possible that the community is not aware that contributors to the project are working on such large features before they land in the repo?

The problem is compounded by the fact that the two-weeks rule makes it possible to merge such a large feature somehow easily.

#  Development Process Transparency

Given the context, I'm advocating for more transparency. Fortunately, I'm not alone thinking that what happened is abnormal. I think that significant features should be introduced and discussed openly before the work happens.

OCIO is used virtually by all the VFX industry DCCs, commercial or open-source, and as such is a critical piece of software for the industry and its community. It, thus, deserves uttermost clarity in its development process. I certainly feel abnormal that some of the key contributors have been developing a significant feature in the shadows, pushed it to the light and can have it merged within two weeks. I searched the mailing list, the issues, the TSC notes, the Working Group notes and I could not see anything pertaining to that and even if I failed to find it, it should have been easy in the first place.

Don't get me wrong here, like everybody else, I sincerely appreciate the work Doug, Patrick and Bernard have been doing, OCIO resurrected from the ashes thanks to their efforts. 🎊 

Back to the issue at hand: What can we do to avoid such a situation again?

I'm hoping that the process we define here can be applied to other ASWF projects because OCIO is not the only one affected by the sound of it.

# Roadmap

The immediate answer to the problem is that any significant feature on a critical and stable project like OCIO should be discussed beforehand in an issue or, at the very least, during the early stage of the feature development. Sometimes, a small toy project, a test prototype can be turned into a feature and the idea here is not to prevent such initiatives, quite the opposite.

Creating an issue for the feature gives time to the community to discuss it and decide whether it is appropriate for the project. It is also important because it avoids potentially wasted time and energy: Imagine a contributor spending 2 or 3 weeks producing 20K lines of fine code just to get his work rejected because it is deemed out-of-scope.

The other part of the answer is that anything significant on the project should be tied to a roadmap, I know that roadmaps are two-edged swords, especially when stakeholders have expectations about things being delivered on time. I'm not looking for that but more for a view that gives a high-level summary of what is being worked on the project, akin to an Agile view where one can only see the Epics for the project.

An example I gave is Epic Games' Trello Board: https://trello.com/b/TTAVI7Ny/ue4-roadmap

![image](https://user-images.githubusercontent.com/99779/91521482-d8f56000-e94b-11ea-817b-2c0482269f8a.png)

I will shut-up and let people chime in!

Bests,

Thomas

**Comment by doug-walker** [comment]
Hi Thomas,

Thanks for creating this issue for discussion of the development process.  I'd like to refresh people's memories of the development process we've been following for the OCIO v2 project.

We started with a written proposal for OCIO v2 that was presented to the community along with a presentation at the BoF at SIGGRAPH 2017.

We also have an OCIO v2 working group.  This is a Google group that anyone is welcome to join:
https://groups.google.com/g/ocio-v2

The group meets regularly on Zoom and we discuss all new features at working group meetings.  

I also create Issues for significant new features on GitHub to capture additional discussions from people that are unable to attend the meetings.

The code review process on GitHub requires two approvals from a specific set of OCIO committers.  We agreed that an Autodesk person would not approve another Autodesk person's PR until after two weeks had elapsed with no objections to the PR from the community (the "two-week rule").  This means that the community always has a substantial window of time to comment on work before it is merged.  Nothing is merged if there are pending objections.

I disagree with your characterization of the two-week rule as an "easy" process.  Michael is basically the only one that takes time to review our PRs.  This means that we have to let the two-week clock run out on basically everything we do before merging it.  The end result of that is that we cannot work in small PRs (as we do internally at Autodesk), because the two week delay on each one would make it impractical.  

For the specific feature you are commenting on, we did discuss it at a working group meeting.  However, there was a delay between the meeting and when I created the issue on GitHub, which was only a day or two before the PR.  We've had a huge number of tasks running in parallel as SIGGRAPH approached (writing the DigiPro paper, working on the new documentation, preparing presentations, working on other features, etc.).  I apologize for not posting the issue for public comment sooner.  That said, no one has posted any comments on the actual feature itself within the issue in the two weeks since it was created.

So we did not perfectly follow our usual process in this instance.  In most three year projects, there are going to be a few items that do not go exactly as desired.  It does not necessarily mean that the process is broken and I'm hoping that one minor instance will not overshadow our entire track record.  But of course, I am very happy to revisit our development process and to discuss what improvements could be made.  

Thank you for sharing the example of the Unreal Engine Trello roadmap.  While it is certainly an admirable example, we would need to come up with an approach that is feasible given the level of activity on OCIO.  Obviously there is a difference in scale between Unreal Engine and OCIO of several orders of magnitude. 

We have started using GitHub Projects on OCIO for the Python testing and Documentation sub-projects, but there is some overhead to keep those up to date and it does not always happen.  Also, they do not give an accurate indication of how close the project is to completion since we tend to work Agile and only enter projects as we get to them rather than trying to create tasks for the entire project up front.

At Autodesk, we do use tools such as Trello and JIRA to plan and communicate feature development work, however we also have a person (the Scrum master) that spends a good part of their time managing the process and we have dedicated product designers and QA people that help keep these updated so that the whole process does not fall on the developers.  

Regarding the Roadmap, we gave a roadmap presentation last year at SIGGRAPH and at the ASWF plenary meeting in January and also wrote a "State of the Project" blog-post on the ASWF site in the spring.  So we have been trying to keep people informed about the progress and provide guidance about when v2 would be ready.  Fortunately, our estimates have been accurate and we have reached Feature Complete when we said we would.  That said, there have been a lot of late nights and weekends spent trying to keep us on that schedule and not everything always goes perfectly.   :)

So in summary, I hope that people do realize that we have been making a considerable effort to be transparent and to communicate our roadmap.  Of course, we also believe in continuous improvement and we're happy to have a discussion about what ideas people have to improve the process.

Doug

**Comment by lgritz** [comment]
Well handled response, Doug.

I'm not super plugged into the OCIO dev these days, but I recognize characteristics of other projects I work on in the description above. Two quotes resonated for me especially:

> Michael is basically the only one that takes time to review our PRs. This means that we have to let the two-week clock run out on basically everything we do before merging it. The end result of that is that we cannot work in small PRs (as we do internally at Autodesk), because the two week delay on each one would make it impractical.

> At Autodesk ... we also have a person (the Scrum master) that spends a good part of their time managing the process and we have dedicated product designers and QA people that help keep these updated so that the whole process does not fall on the developers.

What I'm about to say is not in any way a refutation of @KelSolaar's frustration, which I also understand, but more of a plea not only for OCIO but also on behalf of all the other projects in our ecosystem that are critical infrastructure for a huge number of people and projects, but which have severely undersized development and other resources. I face these same problems every day on OIIO and OSL.

We need more people actively participating. Being the front line for answering issues. Reviewing PRs in a timely fashion. Doing the development on all the smaller topics (which, if the core developers have to do them, detract from their tackling major features at a pace and workstyle that seems more transparent and considered). Managing releases and schedule. Being the scrum master or whatever.

A dozen dedicated stakeholders participating fully and consistently -- even if very much part time -- would drastically change the lives and productivity of the couple of full-time developers on a project in a way that would help to directly address the concerns expressed above. Having several people at the ready who could help turn around most PRs in 2-3 days consistently instead of almost always running out the 2 week clock would drastically change the dynamic and allow for smaller, more incremental PRs and fewer enormous changes.

**Comment by KelSolaar** [comment]
Hi Doug,

Thank you for your answer.

Many transparent pieces scattered everywhere does not necessarily directly correlate with a process that is perceived as fully transparent.

I will explain where I'm coming from in the hope it clarifies where I'm going to. People external to the project and reading the OP might be wondering what that guy with null code contribution is talking about.

_Where I'm coming from:_

So I have been like many here using OCIO for many years, lamenting about its slow decay. I was at the Siggraph 2017 OCIO BoF and also at the first OCIO 2 Working Group Meeting at the JW Marriott (where I said that I could not contribute to OCIO directly but I would continue to author and maintain the ACES Config). This was a great meeting btw, the Phoenix rose from the ashes that day and filled a lot of us with hopes and excitement, so thank you again Doug! I attend most of the Working Group Meetings and read all the notes you send. I watch the repo and, at the very least diagonally read, every single issue or PR. I participate in some issue discussions and open some from time to time. Every week, I religiously read the TSC Meeting notes (except that of the Documentation one though, sorry Carol, I just complain about docs :)). I'm on the mailing list, not so active and on Slack, quite active. I try to read or watch all the presentations, slide decks, etc... So I think it is fair to say that I'm generally well aware of the state of things.

_Where I'm going to:_

If I felt that #1100 and #1102 came as a surprise, even to some TSC members, how do you think it might be perceived by people not following the project as I do?

The issue here might be simply that there is an implicit assumption that everybody has the same knowledge about the state of the project and is up-to-speed but this is not true. It requires a considerable amount of effort to be up-to-date and assemble all the scattered pieces of the puzzle. So yes, while abstracting #1100 (which was akin to open a pandora box without the scope you gave following my initial reply), you have always been transparent but it, unfortunately, does not mean that there is no general opacity in the development process.

I think we need to come up with something that helps people:

- Understand where the project is at
- Understand where the project is going
- Understand the process adopted to make that happen

And most importantly, this needs to be easy to find from the README and from the Homepage. Ideally, someones outside should get up-to-speed via those means.

Cheers,

Thomas

PS: I'm taking note for Colour while writing that, otherwise people will point their fingers at me.

**Comment by KelSolaar** [comment]
@lgritz : I think we just cross-posted! :)

**Comment by lgritz** [comment]
Yeah. Complementary takes on it, I think.

**Comment by KelSolaar** [comment]
Yes exactly what I thought when reading your post :)

**Comment by michdolan** [comment]
Thank you @KelSolaar for the feedback and raising good points for community discussion, and thanks @doug-walker for your well thought out response. I recognize that there are plenty of areas where we can still improve as a project, and the TSC is certainly open to implementing these. I also would encourage that anyone who shares these or other concerns to please follow Thomas's example by bringing them up in a community forum where they can be addressed.

I think one practical change we should implement to help with this concern is to merge the v2 working group with the TSC, and allow more regular proposal/discussion on in-development features in that forum. Our move to the ASWF came at a busy time in project development, so we have been largely using the TSC meetings to discuss administration and process, and the v2 working group to discuss design and features. Most of the administrative requirements are in place now, so we have time to focus on functionality.

We can return the TSC meeting duration to an hour and also move it to a later time in the day if that helps be more global time-zone friendly.

On the point of coming up with something with more detail around the development process, we could add a `PROCESS.md` to the repo root to compliment `GOVERNANCE.md` and `CONTRIBUTING.md` and more clearly outline the feature/change design and development process. And this is something that can probably be brought back to the ASWF for wider discussion too. The contents for that document can be discussed here and document the current Autodesk process, with any adjustments that the community feels would improve project transparency.

**Comment by carolalynn22** [comment]
All of this discussion is great! I really like the idea of merging back the TSC and working group. A thing I've noticed that we all do (myself very much included) is be quite vocal about something in a working group meeting, but silent on the issue when raised in GitHub, which contributes as well to the lag Doug describes quite accurately in PR review. 

I think finding a way to highlight these issues so they are easy to find, track and contribute to, but at the same time not too much work to maintain/update, will be key. 

As Doug also mentions, the role of a scrum master / project manager is a full time, key role in all of our development processes at our respective companies, but one that isn't visible in the same way in most open source development. Wondering that as we try and tackle this issue in OCIO, if there is also a larger issue to be raised/discussed at the ASWF level as to how to integrate/encourage open source participation in a project management type role.

**Comment by KelSolaar** [comment]
> I really like the idea of merging back the TSC and working group.

With the hope that they can account for the TZ of the Hobbits living in The Shire :)

**Comment by michdolan** [comment]
> With the hope that they can account for the TZ of the Hobbits living in The Shire :)

We could adopt the somewhat standard AWSF meeting time slot (1pm PDT/4pm EST/9pm BST/9am NZST)

Also, stumbled on this repo from Adobe which has some good ideas in terms of documenting process: https://github.com/adobe/open-development-template

**Comment by doug-walker** [comment]
It's been a busy week for many of us and I really appreciate people taking time to express their thoughts and ideas.

@michdolan, I fully support the concrete proposals you made.  

The OCIO v2 working group actually predated the formation of the OCIO TSC.  And as the project was going through the graduation process from incubating to adopted status there was so much work totally separate from v2 that it made sense to have separate meetings.  Indeed, the work Michael lead on the ASWF-related tasks was a huge benefit to the Autodesk team and allowed us to focus on the v2 features (rather than the CI system, CII badge, license scan, etc.).  

But now that the project has obtained ASWF adopted status and v2 is nearing release, I agree it makes sense to wind down the v2 working group and handle those discussions in the TSC meetings.  Hopefully that will provide some of the added transparency that Thomas and his friends are looking for.

(I will note that it has been a slightly different set of people in the two groups.  But there is so much overlap that I don't think that should be a problem.)

Let's discuss the TSC time change at the Monday meeting.  As long as it's ok with our London-based friends, it's fine with Patrick and I.

As Carol and Larry point out, the discussion raises open source project management issues that go beyond OCIO.  I totally agree with Larry's point that having a larger group of people that are actively helping to get work done in GitHub (reviewing PRs, helping manage issues, etc.) would have a dramatic benefit.  And as Carol wrote, these do not necessarily need to be developers, more project management help would also target the issue Thomas has brought up.

At the end of the day, it seems we need to encourage more people to make the case with their employer that this would be a good use of their time.  These tasks involve real work, and experience seems to have shown that the number of people willing to do them on nights and weekends is very limited (although not surprisingly, the people posting here are all notable exceptions!).  Additional recommendations along these lines from the ASWF would indeed be very welcome so that we could benefit from that wealth of experience.  Patrick and I would be very happy to participate in any discussions and share our perspective on the past three years of OCIO development.

Doug
